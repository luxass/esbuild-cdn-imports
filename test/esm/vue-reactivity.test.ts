import { assert, describe, expect, it } from "vitest";
import { buildWithCDN } from "../__utils";

describe("esm: @vue/reactivity", () => {
  it("resolve @vue/reactivity", async () => {
    const { result } = await buildWithCDN(
      "test/fixtures/vue-reactivity",
      {
        cdn: "esm",
        fileName: "vue-reactivity.ts",
      },
    );

    assert(result.outputFiles, "outputFiles should be defined");

    const outputText = result.outputFiles[0].text;

    expect(outputText).toMatch(
      /^\/\/ cdn-imports:https:\/\/esm.sh\/@vue\/reactivity/gm,
    );

    const matchedImports = result.outputFiles[0].text.match(
      /^\/\/ cdn-imports:https:\/\/esm.sh\/@vue\//gm,
    );

    expect(matchedImports).toHaveLength(2);
  });

  it("resolve @vue/reactivity@3.1.5", async () => {
    const { result } = await buildWithCDN(
      "test/fixtures/vue-reactivity",
      {
        cdn: "esm",
        fileName: "vue-reactivity.ts",
        versions: {
          "@vue/reactivity": "3.1.5",
        },
      },
    );

    assert(result.outputFiles, "outputFiles should be defined");
    const outputText = result.outputFiles[0].text;

    expect(outputText).toMatchSnapshot();

    expect(outputText).toMatch(
      /^\/\/ cdn-imports:https:\/\/esm.sh\/@vue\/reactivity@3\.1\.5/gm,
    );

    const matchedImports = result.outputFiles[0].text.match(
      /^\/\/ cdn-imports:https:\/\/esm.sh\/@vue\//gm,
    );

    expect(matchedImports).toHaveLength(2);
  });
});
