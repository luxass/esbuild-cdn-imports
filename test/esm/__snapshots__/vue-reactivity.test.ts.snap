// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`esm: @vue/reactivity > resolve @vue/reactivity@3.1.5 1`] = `
"// cdn-imports:https://esm.sh/@vue/shared@3.1.5/es2022/shared.mjs
function l(e, t) {
  let n = /* @__PURE__ */ Object.create(null), o = e.split(",");
  for (let r = 0; r < o.length; r++) n[o[r]] = true;
  return t ? (r) => !!n[r.toLowerCase()] : (r) => !!n[r];
}
var N = "Infinity,undefined,NaN,isFinite,isNaN,parseFloat,parseInt,decodeURI,decodeURIComponent,encodeURI,encodeURIComponent,Math,Number,Date,Array,Object,Boolean,String,RegExp,Map,Set,JSON,Intl,BigInt";
var ee = l(N);
var k = "itemscope,allowfullscreen,formnovalidate,ismap,nomodule,novalidate,readonly";
var ne = l(k);
var oe = l(k + ",async,autofocus,autoplay,controls,default,defer,disabled,hidden,loop,open,required,reversed,scoped,seamless,checked,muted,multiple,selected");
var R = l("animation-iteration-count,border-image-outset,border-image-slice,border-image-width,box-flex,box-flex-group,box-ordinal-group,column-count,columns,flex,flex-grow,flex-positive,flex-shrink,flex-negative,flex-order,grid-row,grid-row-end,grid-row-span,grid-row-start,grid-column,grid-column-end,grid-column-span,grid-column-start,font-weight,line-clamp,line-height,opacity,order,orphans,tab-size,widows,z-index,zoom,fill-opacity,flood-opacity,stop-opacity,stroke-dasharray,stroke-dashoffset,stroke-miterlimit,stroke-opacity,stroke-width");
var ie = l("accept,accept-charset,accesskey,action,align,allow,alt,async,autocapitalize,autocomplete,autofocus,autoplay,background,bgcolor,border,buffered,capture,challenge,charset,checked,cite,class,code,codebase,color,cols,colspan,content,contenteditable,contextmenu,controls,coords,crossorigin,csp,data,datetime,decoding,default,defer,dir,dirname,disabled,download,draggable,dropzone,enctype,enterkeyhint,for,form,formaction,formenctype,formmethod,formnovalidate,formtarget,headers,height,hidden,high,href,hreflang,http-equiv,icon,id,importance,integrity,ismap,itemprop,keytype,kind,label,lang,language,loading,list,loop,low,manifest,max,maxlength,minlength,media,min,multiple,muted,name,novalidate,open,optimum,pattern,ping,placeholder,poster,preload,radiogroup,readonly,referrerpolicy,rel,required,reversed,rows,rowspan,sandbox,scope,scoped,selected,shape,size,sizes,slot,span,spellcheck,src,srcdoc,srclang,srcset,start,step,style,summary,tabindex,target,title,translate,type,usemap,value,width,wrap");
var _ = "html,body,base,head,link,meta,style,title,address,article,aside,footer,header,h1,h2,h3,h4,h5,h6,hgroup,nav,section,div,dd,dl,dt,figcaption,figure,picture,hr,img,li,main,ol,p,pre,ul,a,b,abbr,bdi,bdo,br,cite,code,data,dfn,em,i,kbd,mark,q,rp,rt,rtc,ruby,s,samp,small,span,strong,sub,sup,time,u,var,wbr,area,audio,map,track,video,embed,object,param,source,canvas,script,noscript,del,ins,caption,col,colgroup,table,thead,tbody,td,th,tr,button,datalist,fieldset,form,input,label,legend,meter,optgroup,option,output,progress,select,textarea,details,dialog,menu,summary,template,blockquote,iframe,tfoot";
var I = "svg,animate,animateMotion,animateTransform,circle,clipPath,color-profile,defs,desc,discard,ellipse,feBlend,feColorMatrix,feComponentTransfer,feComposite,feConvolveMatrix,feDiffuseLighting,feDisplacementMap,feDistanceLight,feDropShadow,feFlood,feFuncA,feFuncB,feFuncG,feFuncR,feGaussianBlur,feImage,feMerge,feMergeNode,feMorphology,feOffset,fePointLight,feSpecularLighting,feSpotLight,feTile,feTurbulence,filter,foreignObject,g,hatch,hatchpath,image,line,linearGradient,marker,mask,mesh,meshgradient,meshpatch,meshrow,metadata,mpath,path,pattern,polygon,polyline,radialGradient,rect,set,solidcolor,stop,switch,symbol,text,textPath,title,tspan,unknown,use,view";
var F = "area,base,br,col,embed,hr,img,input,link,meta,param,source,track,wbr";
var ce = l(_);
var le = l(I);
var pe = l(F);
var Oe = Object.assign;
var q = Object.prototype.hasOwnProperty;
var xe = (e, t) => q.call(e, t);
var f = Array.isArray;
var $ = (e) => g(e) === "[object Map]";
var h = (e) => typeof e == "string";
var ke = (e) => typeof e == "symbol";
var p = (e) => e !== null && typeof e == "object";
var K = Object.prototype.toString;
var g = (e) => K.call(e);
var Ne = (e) => g(e).slice(8, -1);
var Me = (e) => h(e) && e !== "NaN" && e[0] !== "-" && "" + parseInt(e, 10) === e;
var Re = l(",key,ref,onVnodeBeforeMount,onVnodeMounted,onVnodeBeforeUpdate,onVnodeUpdated,onVnodeBeforeUnmount,onVnodeUnmounted");
var m = (e) => {
  let t = /* @__PURE__ */ Object.create(null);
  return (n) => t[n] || (t[n] = e(n));
};
var W = /-(\\w)/g;
var Ce = m((e) => e.replace(W, (t, n) => n ? n.toUpperCase() : ""));
var J = /\\B([A-Z])/g;
var X = m((e) => e.replace(J, "-$1").toLowerCase());
var Z = m((e) => e.charAt(0).toUpperCase() + e.slice(1));
var De = m((e) => e ? \`on\${Z(e)}\` : "");
var Le = (e, t) => e !== t && (e === e || t === t);

// cdn-imports:https://esm.sh/@vue/reactivity@3.1.5/es2022/reactivity.mjs
var z = /* @__PURE__ */ new WeakMap();
var g2;
var v = Symbol("");
var $2 = Symbol("");
var E2 = true;
var U = [];
function me2() {
  U.push(E2), E2 = false;
}
function oe2() {
  let e = U.pop();
  E2 = e === void 0 ? true : e;
}
function f2(e, t, n) {
  if (!E2 || g2 === void 0) return;
  let r = z.get(e);
  r || z.set(e, r = /* @__PURE__ */ new Map());
  let s = r.get(n);
  s || r.set(n, s = /* @__PURE__ */ new Set()), s.has(g2) || (s.add(g2), g2.deps.push(s));
}
function h2(e, t, n, r, s, o) {
  let i = z.get(e);
  if (!i) return;
  let a2 = /* @__PURE__ */ new Set(), u = (l2) => {
    l2 && l2.forEach((d) => {
      (d !== g2 || d.allowRecurse) && a2.add(d);
    });
  };
  if (t === "clear") i.forEach(u);
  else if (n === "length" && f(e)) i.forEach((l2, d) => {
    (d === "length" || d >= r) && u(l2);
  });
  else switch (n !== void 0 && u(i.get(n)), t) {
    case "add":
      f(e) ? Me(n) && u(i.get("length")) : (u(i.get(v)), $(e) && u(i.get($2)));
      break;
    case "delete":
      f(e) || (u(i.get(v)), $(e) && u(i.get($2)));
      break;
    case "set":
      $(e) && u(i.get(v));
      break;
  }
  let R2 = (l2) => {
    l2.options.scheduler ? l2.options.scheduler(l2) : l2();
  };
  a2.forEach(R2);
}
var Ie = l("__proto__,__v_isRef,__isVue");
var ie2 = new Set(Object.getOwnPropertyNames(Symbol).map((e) => Symbol[e]).filter(ke));
var Ke = O();
var Oe2 = O(false, true);
var Ve = O(true);
var xe2 = O(true, true);
var X2 = De2();
function De2() {
  let e = {};
  return ["includes", "indexOf", "lastIndexOf"].forEach((t) => {
    e[t] = function(...n) {
      let r = c(this);
      for (let o = 0, i = this.length; o < i; o++) f2(r, "get", o + "");
      let s = r[t](...n);
      return s === -1 || s === false ? r[t](...n.map(c)) : s;
    };
  }), ["push", "pop", "shift", "unshift", "splice"].forEach((t) => {
    e[t] = function(...n) {
      me2();
      let r = c(this)[t].apply(this, n);
      return oe2(), r;
    };
  }), e;
}
function O(e = false, t = false) {
  return function(r, s, o) {
    if (s === "__v_isReactive") return !e;
    if (s === "__v_isReadonly") return e;
    if (s === "__v_raw" && o === (e ? t ? de : he : t ? fe : le2).get(r)) return r;
    let i = f(r);
    if (!e && i && xe(X2, s)) return Reflect.get(X2, s, o);
    let a2 = Reflect.get(r, s, o);
    return (ke(s) ? ie2.has(s) : Ie(s)) || (e || f2(r, "get", s), t) ? a2 : w(a2) ? !i || !Me(s) ? a2.value : a2 : p(a2) ? e ? _e(a2) : q2(a2) : a2;
  };
}
var Pe = ce2();
var ze = ce2(true);
function ce2(e = false) {
  return function(n, r, s, o) {
    let i = n[r];
    if (!e && (s = c(s), i = c(i), !f(n) && w(i) && !w(s))) return i.value = s, true;
    let a2 = f(n) && Me(r) ? Number(r) < n.length : xe(n, r), u = Reflect.set(n, r, s, o);
    return n === c(o) && (a2 ? Le(s, i) && h2(n, "set", r, s, i) : h2(n, "add", r, s)), u;
  };
}
function $e(e, t) {
  let n = xe(e, t), r = e[t], s = Reflect.deleteProperty(e, t);
  return s && n && h2(e, "delete", t, void 0, r), s;
}
function He(e, t) {
  let n = Reflect.has(e, t);
  return (!ke(t) || !ie2.has(t)) && f2(e, "has", t), n;
}
function Ae(e) {
  return f2(e, "iterate", f(e) ? "length" : v), Reflect.ownKeys(e);
}
var ae = { get: Ke, set: Pe, deleteProperty: $e, has: He, ownKeys: Ae };
var ue = { get: Ve, set(e, t) {
  return true;
}, deleteProperty(e, t) {
  return true;
} };
var We = Oe({}, ae, { get: Oe2, set: ze });
var Ge = Oe({}, ue, { get: xe2 });
var B = (e) => p(e) ? q2(e) : e;
var F2 = (e) => p(e) ? _e(e) : e;
var J2 = (e) => e;
var V = (e) => Reflect.getPrototypeOf(e);
function b(e, t, n = false, r = false) {
  e = e.__v_raw;
  let s = c(e), o = c(t);
  t !== o && !n && f2(s, "get", t), !n && f2(s, "get", o);
  let { has: i } = V(s), a2 = r ? J2 : n ? F2 : B;
  if (i.call(s, t)) return a2(e.get(t));
  if (i.call(s, o)) return a2(e.get(o));
  e !== s && e.get(t);
}
function M(e, t = false) {
  let n = this.__v_raw, r = c(n), s = c(e);
  return e !== s && !t && f2(r, "has", e), !t && f2(r, "has", s), e === s ? n.has(e) : n.has(e) || n.has(s);
}
function S(e, t = false) {
  return e = e.__v_raw, !t && f2(c(e), "iterate", v), Reflect.get(e, "size", e);
}
function Z2(e) {
  e = c(e);
  let t = c(this);
  return V(t).has.call(t, e) || (t.add(e), h2(t, "add", e, e)), this;
}
function k2(e, t) {
  t = c(t);
  let n = c(this), { has: r, get: s } = V(n), o = r.call(n, e);
  o || (e = c(e), o = r.call(n, e));
  let i = s.call(n, e);
  return n.set(e, t), o ? Le(t, i) && h2(n, "set", e, t, i) : h2(n, "add", e, t), this;
}
function ee2(e) {
  let t = c(this), { has: n, get: r } = V(t), s = n.call(t, e);
  s || (e = c(e), s = n.call(t, e));
  let o = r ? r.call(t, e) : void 0, i = t.delete(e);
  return s && h2(t, "delete", e, void 0, o), i;
}
function te() {
  let e = c(this), t = e.size !== 0, n = void 0, r = e.clear();
  return t && h2(e, "clear", void 0, void 0, n), r;
}
function m2(e, t) {
  return function(r, s) {
    let o = this, i = o.__v_raw, a2 = c(i), u = t ? J2 : e ? F2 : B;
    return !e && f2(a2, "iterate", v), i.forEach((R2, l2) => r.call(s, u(R2), u(l2), o));
  };
}
function T(e, t, n) {
  return function(...r) {
    let s = this.__v_raw, o = c(s), i = $(o), a2 = e === "entries" || e === Symbol.iterator && i, u = e === "keys" && i, R2 = s[e](...r), l2 = n ? J2 : t ? F2 : B;
    return !t && f2(o, "iterate", u ? $2 : v), { next() {
      let { value: d, done: P } = R2.next();
      return P ? { value: d, done: P } : { value: a2 ? [l2(d[0]), l2(d[1])] : l2(d), done: P };
    }, [Symbol.iterator]() {
      return this;
    } };
  };
}
function _2(e) {
  return function(...t) {
    return e === "delete" ? false : this;
  };
}
function je2() {
  let e = { get(o) {
    return b(this, o);
  }, get size() {
    return S(this);
  }, has: M, add: Z2, set: k2, delete: ee2, clear: te, forEach: m2(false, false) }, t = { get(o) {
    return b(this, o, false, true);
  }, get size() {
    return S(this);
  }, has: M, add: Z2, set: k2, delete: ee2, clear: te, forEach: m2(false, true) }, n = { get(o) {
    return b(this, o, true);
  }, get size() {
    return S(this, true);
  }, has(o) {
    return M.call(this, o, true);
  }, add: _2("add"), set: _2("set"), delete: _2("delete"), clear: _2("clear"), forEach: m2(true, false) }, r = { get(o) {
    return b(this, o, true, true);
  }, get size() {
    return S(this, true);
  }, has(o) {
    return M.call(this, o, true);
  }, add: _2("add"), set: _2("set"), delete: _2("delete"), clear: _2("clear"), forEach: m2(true, true) };
  return ["keys", "values", "entries", Symbol.iterator].forEach((o) => {
    e[o] = T(o, false, false), n[o] = T(o, true, false), t[o] = T(o, false, true), r[o] = T(o, true, true);
  }), [e, n, t, r];
}
var [Ye, Ce2, Ue, Be] = je2();
function x(e, t) {
  let n = t ? e ? Be : Ue : e ? Ce2 : Ye;
  return (r, s, o) => s === "__v_isReactive" ? !e : s === "__v_isReadonly" ? e : s === "__v_raw" ? r : Reflect.get(xe(n, s) && s in r ? n : r, s, o);
}
var Fe = { get: x(false, false) };
var Je = { get: x(false, true) };
var qe = { get: x(true, false) };
var Le2 = { get: x(true, true) };
var le2 = /* @__PURE__ */ new WeakMap();
var fe = /* @__PURE__ */ new WeakMap();
var he = /* @__PURE__ */ new WeakMap();
var de = /* @__PURE__ */ new WeakMap();
function Qe(e) {
  switch (e) {
    case "Object":
    case "Array":
      return 1;
    case "Map":
    case "Set":
    case "WeakMap":
    case "WeakSet":
      return 2;
    default:
      return 0;
  }
}
function Xe(e) {
  return e.__v_skip || !Object.isExtensible(e) ? 0 : Qe(Ne(e));
}
function q2(e) {
  return e && e.__v_isReadonly ? e : D(e, false, ae, Fe, le2);
}
function _e(e) {
  return D(e, true, ue, qe, he);
}
function D(e, t, n, r, s) {
  if (!p(e) || e.__v_raw && !(t && e.__v_isReactive)) return e;
  let o = s.get(e);
  if (o) return o;
  let i = Xe(e);
  if (i === 0) return e;
  let a2 = new Proxy(e, i === 2 ? r : n);
  return s.set(e, a2), a2;
}
function c(e) {
  return e && c(e.__v_raw) || e;
}
var ne2 = (e) => p(e) ? q2(e) : e;
function w(e) {
  return !!(e && e.__v_isRef === true);
}
function at(e) {
  return we(e);
}
var H = class {
  constructor(t, n = false) {
    this._shallow = n, this.__v_isRef = true, this._rawValue = n ? t : c(t), this._value = n ? t : ne2(t);
  }
  get value() {
    return f2(c(this), "get", "value"), this._value;
  }
  set value(t) {
    t = this._shallow ? t : c(t), Le(t, this._rawValue) && (this._rawValue = t, this._value = this._shallow ? t : ne2(t), h2(c(this), "set", "value", t));
  }
};
function we(e, t = false) {
  return w(e) ? e : new H(e, t);
}

// .vitest-testdirs/vitest-vue-reactivity-esm-vue-reactivity-resolve-vue-reactivity-3-1-5/src/vue-reactivity.ts
var a = at(1);
"
`;
